scripts/genVis.py

```diff
@@ -1,58 +1,140 @@
+                                                    'Specify dimension (2 or 3) as argument. Default is 2.')
+parser.add_argument('-c', '--handleExternalClusterFile', action='store_true', default=False,
+                    help='Call to handle an external cluster file.')
+parser.add_argument('-n', '--plotTitle', default='', nargs='*', help='Write a plot title.')
 
-parser.add_option('-p','--pseudoTimeFile',default='',type=str,
-                  help='Specify pseudoTimeFile file name')
+# Parse arguments
+args = parser.parse_args()
+path = args.pathToFiles
+inFile = path + "/expressiondata.csv"
+timeFile = path + "/pseudotime.csv"
+pca_flag = args.pca is not None
+tsne_flag = args.tsne is not None
+umap_flag = args.umap is not None
+clusterFile = args.pathToFiles + "/clusterids.csv"
 
-parser.add_option('-t','--tsne',action='store_true',default=False,
-                  help='Visualized tsne instead of default PCA')
+# Do PCA, tSNE, UMAP
+DF = pd.read_csv(inFile, sep=',', index_col=0)
+Cells = DF.T.values
+DRDF = pd.DataFrame(index=pd.Index(list(DF.columns)))
+if pca_flag:
+    if len(args.pca) == 0:
+        pca_dim = 2
+    else:
+        pca_dim = int(args.pca[0])
+    pca_embed = PCA(n_components=pca_dim).fit_transform(Cells)
+    DRDF['PCA1'] = pca_embed[:, 0]
+    DRDF['PCA2'] = pca_embed[:, 1]
+    if pca_dim == 3:
+        DRDF['PCA3'] = pca_embed[:, 2]
+if tsne_flag:
+    if len(args.tsne) == 0:
+        tsne_dim = 2
+    else:
+        tsne_dim = int(args.tsne[0])
+    tsne_embed = TSNE(n_components=tsne_dim).fit_transform(Cells)
+    DRDF['tsne1'] = tsne_embed[:, 0]
+    DRDF['tsne2'] = tsne_embed[:, 1]
+    if tsne_dim == 3:
+        DRDF['tsne3'] = tsne_embed[:, 2]
+if umap_flag:
+    if len(args.umap) == 0:
+        umap_dim = 2
+    else:
+        umap_dim = int(args.umap[0])
+    umap_embed = umap.UMAP(n_components=umap_dim).fit_transform(Cells)
+    DRDF['umap1'] = umap_embed[:, 0]
+    DRDF['umap2'] = umap_embed[:, 1]
+    if umap_dim == 3:
+        DRDF['umap3'] = umap_embed[:, 2]
 
-(opts, args) = parser.parse_args()
-inFile = opts.inFile
-tsne_flag = opts.tsne
 
+# Prepare time-dependent color scheme
+ptDF = pd.read_csv(timeFile, sep=',', index_col=0)
+color_scale = max(ptDF["Time"])
+colors_raw = [h.split('_')[1] for h in DF.columns]
+colors_raw = [int(i) for i in colors_raw]
+colors = [x / color_scale for x in colors_raw]
+experiments = set([h.split('_')[0] for h in DF.columns])
+DRDF['time'] = colors
 
+# Prepare cluster-dependent color scheme
+if args.handleExternalClusterFile:
+    CF = pd.read_csv(clusterFile, sep=',', index_col=0)
+    cluster_colors_raw = CF['cl'].tolist()
+    cluster_color_scale = max(CF['cl'])
+    cluster_colors = [y / color_scale for y in cluster_colors_raw]
+else:
+    cluster_colors = list(itertools.repeat(.5, len(DF.columns)))
 
+# Write dimensionality reduction data to text file
+DRDF.to_csv(inFile + '_dimred.txt')
```

Comment:
Don't you need different files if the user passes more than one of the pca, tsne, and umap options?

Reply:
I kept this feature from the original version of the `genVis.py` script. All the dimensional reduction data for each cell is stored in a row of the same Dimensionality Reduction Data Frame (DRDF). All this line does is write the DRDF to a single file so that the numerical data from the dimensionality reduction(s) can be accessed at a later time if need be. Would you perhaps prefer that the data be saved as a `.tsv` or a `.csv` instead?